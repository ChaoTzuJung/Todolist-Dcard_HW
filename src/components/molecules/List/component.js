import React, { Fragment } from 'react';
import classnames from 'classnames';

import Checkbox from 'components/atoms/Checkbox';
import Field from 'components/atoms/Field';
import Icon from 'components/atoms/Icon';

import styles from './index.css';

const List = ({
	// state
	cacheTodo: { message },
	completed,
	// message,
	star,
	edit,

	// props
	className,
	// icon 隔壁的時間
	date,
	// 3個小icon
	deadline,
	file,
	comment,
	addStar,
	onEdit,

	// actions
	handleCheckboxChange,
	handleChange,
}) => (
	<Fragment>
		<div
			className={classnames(styles.list, className, {
				[styles.colorList]: star,
				[styles.squaredBottom]: edit,
			})}
		>
			<div className={styles.left}>
				<div className={styles.top}>
					{edit ? (
						<Field
							className={styles.fieldFix}
							name="edit_todo"
							icon=""
							placeholder="Type Something Here..."
							valid="false"
							color={star ? '#FFF2DC' : '#F2F2F2'}
							border="false"
							value={message}
							onChange={handleChange}
						/>
					) : (
						<Checkbox
							className={classnames({
								[styles.lineThrough]: completed,
							})}
							name={message}
							value={message}
							content={message}
							checked={completed}
							onChange={handleCheckboxChange}
						/>
					)}
				</div>
				{!completed && (
					<div className={styles.down}>
						{deadline && (
							<Fragment>
								<Icon className={styles.iconFix}>date_range</Icon>
								{deadline && <span>{date}</span>}
							</Fragment>
						)}
						{file && (
							<Icon className={styles.iconFix} outlined>
								insert_drive_file
							</Icon>
						)}
						{comment && (
							<Icon className={styles.iconFix} outlined>
								textsms
							</Icon>
						)}
					</div>
				)}
			</div>
			<div className={styles.right}>
				<span onClick={addStar} onKeyPress={addStar} role="button" tabIndex="0">
					<Icon
						className={classnames({
							[styles.star]: star,
						})}
					>
						{star ? 'star' : 'star_border'}
					</Icon>
				</span>
				<span onClick={onEdit} onKeyPress={onEdit} role="button" tabIndex="0">
					<Icon
						className={classnames({
							[styles.edit]: edit,
						})}
						outlined={!edit}
					>
						edit
					</Icon>
				</span>
			</div>
		</div>
	</Fragment>
);

export default List;
