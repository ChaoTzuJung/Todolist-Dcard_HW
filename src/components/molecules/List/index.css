.list {
	display: flex;
	justify-content: space-between;
	width: 620px;
	padding: 24px 32px;
	background-color: #f2f2f2;
	border-radius: 5px;

	& > .left {
		width: 100%;

		& > .top {
			display: flex;
			align-items: center;

			& > .line-through {
				text-decoration: line-through;
				color: #9b9b9b;
			}

			& > .todo-context {
				font-size: 24px;
				margin-left: 16px;
			}

			& > .field-fix {
				width: 100%;
				margin-left: 32px;
			}
		}

		& > .down {
			display: flex;
			align-items: center;
			padding-left: 40px;
			height: calc(100% - 30px);

			& > .icon-fix {
				margin-top: 16px;
				color: #757575;
				font-size: 16px;

				&:not(:first-child) {
					margin-left: 16px;
				}
			}

			& > span {
				margin-top: 16px;
				margin-left: 8px;
				color: #757575;
			}
		}
	}

	& > .right {
		display: flex;

		& > span {
			cursor: pointer;

			& > .star {
				color: #f5a623;
			}

			& > .edit {
				color: #4a90e2;
			}

			& :not(:first-child) {
				margin-left: 32px;
			}
		}
	}
}

.color-list {
	background-color: #fff2dc;
}

.squared-bottom {
	border-radius: 5px 5px 0 0;
}
