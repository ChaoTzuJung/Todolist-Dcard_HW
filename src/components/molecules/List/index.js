import { contain } from 'react-container-helper';

import { firebaseDB } from '../../../../firebase';
import List from './component';

const initState = ({ completed }) => ({
	// controller input
	message: '',
	completed: completed || false,
	// cacheTodo: {
	// 	message: message || '',
	// },
	star: false,
	edit: false,
});

const mapSetStateToProps = (
	{ message, completed, cacheTodo, star, edit },
	{
		className,
		date,
		deadline,
		file,
		comment,
		isNewTodo,
		addStar,
		onEdit,
		handleCheckboxChange,
		...other
	},
	setState,
) => ({
	// state
	message,
	completed,
	cacheTodo,
	star,
	edit,
	// props
	className,
	date,
	deadline,
	file,
	comment,
	addStar,
	onEdit,
	handleCheckboxChange,
	...other,

	handleChange(e) {
		const val = e.target.value;
		console.log(`handleChange 輸入新的message: ${val}`);
		setState(prevState => ({
			...prevState,
			message: val,
		}));
		// () => {
		// 	setState(prevState => ({
		// 		cacheTodo: {
		// 			...prevState.cacheTodo,
		// 			message: prevState.message,
		// 		},
		// 	}));
		// },
	},
});

const setLifecycle = () => ({
	// 每次更新畫面 要 抓 data 的 message 到 cacheTodo 的 state 內
	componentDidMount({ setState, getProps }) {
		const { id, isNewTodo } = getProps();
		console.log('**List** componentDidMount 嘍 !!');
		console.log('**List** id是', id);
		if (!isNewTodo) {
			firebaseDB
				.ref(`todos/${id}`)
				.once('value')
				.then(snapshot => {
					console.log('Cannot read property message of null', snapshot.val());
					const { message } = snapshot.val();
					setState(prevState => ({
						...prevState,
						message,
					}));
					// setState(prevState => ({
					// 	cacheTodo: {
					// 		...prevState.cacheTodo,
					// 		message,
					// 	},
					// }));
				});
		}
	},
});

export default contain(initState, mapSetStateToProps, setLifecycle)(List);
