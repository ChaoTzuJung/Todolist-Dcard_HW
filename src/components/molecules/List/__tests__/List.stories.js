import React from 'react';
import { withKnobs, object, boolean } from '@storybook/addon-knobs/react';
import { storiesOf } from '@storybook/react';

import List from 'components/molecules/List';

const stories = storiesOf('Commons|molecules/List', module);

stories.addDecorator(withKnobs);

stories.add('__interactive', () => (
	<div style={{ margin: '16px' }}>
		<List
			text={object('text', 'type something here...')}
			star={boolean('star', false)}
			edit={boolean('edit', false)}
			checked={boolean('checked', false)}
			date={boolean('date', true)}
			file={boolean('file', true)}
			comment={boolean('comment', true)}
		/>
	</div>
));
