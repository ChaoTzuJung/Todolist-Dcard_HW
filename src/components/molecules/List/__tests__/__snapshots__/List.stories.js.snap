// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`__interactive 1`] = `
<div
  style={
    Object {
      "margin": "16px",
    }
  }
>
  <Contain(List)
    checked={false}
    comment={true}
    date="6/18"
    deadline={true}
    edit={false}
    file={true}
    star={false}
    text="type something here..."
  >
    <List
      AddStar={[Function]}
      OpenEdit={[Function]}
      checked={false}
      comment={true}
      date="6/18"
      deadline={true}
      edit={false}
      file={true}
      handleChange={[Function]}
      onChangeCheckbox={[Function]}
      star={false}
      text="type something here..."
    >
      <div
        className="list"
      >
        <div
          className="left"
        >
          <div
            className="top"
          >
            <Checkbox
              checked={false}
              className=""
              content="type something here..."
              name="todo"
              onChange={[Function]}
            >
              <label
                className="label"
                htmlFor="todo"
              >
                type something here...
                <input
                  checked={false}
                  id="todo"
                  name="todo"
                  onChange={[Function]}
                  type="checkbox"
                />
                <span
                  className="checkbox"
                >
                  <Icon>
                    <i
                      className="icon material-icons"
                    >
                      check
                    </i>
                  </Icon>
                </span>
              </label>
            </Checkbox>
          </div>
          <div
            className="down"
          >
            <Icon
              className="iconFix"
            >
              <i
                className="icon iconFix material-icons"
              >
                date_range
              </i>
            </Icon>
            <span>
              6/18
            </span>
            <Icon
              className="iconFix"
              outlined={true}
            >
              <i
                className="icon iconFix material-icons-outlined"
              >
                insert_drive_file
              </i>
            </Icon>
            <Icon
              className="iconFix"
              outlined={true}
            >
              <i
                className="icon iconFix material-icons-outlined"
              >
                textsms
              </i>
            </Icon>
          </div>
        </div>
        <div
          className="right"
        >
          <span
            role="button"
            tabIndex="0"
          >
            <Icon
              className=""
            >
              <i
                className="icon material-icons"
              >
                star_border
              </i>
            </Icon>
          </span>
          <span
            role="button"
            tabIndex="0"
          >
            <Icon
              className=""
              outlined={true}
            >
              <i
                className="icon material-icons-outlined"
              >
                edit
              </i>
            </Icon>
          </span>
        </div>
      </div>
    </List>
  </Contain(List)>
</div>
`;
